File: integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/BookDTO.java
package io.quarkus.it.mongodb.panache;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonFormat;

import io.quarkus.it.mongodb.panache.book.BookDetail;

/**
 * The IT uses a DTO and not directly the Book object because it should avoid the usage of ObjectId.
 */
public class BookDTO {
    private String title;
    private String author;
    private String id;
    private String transientDescription;
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
    private Date creationDate;

    private List<String> categories = new ArrayList<>();

    private BookDetail details;

    // will only be used in case of projection
    private Integer rating;

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public Integer getRating() {
        return rating;
    }

    public void setRating(Integer rating) {
        this.rating = rating;
    }

    public String getTitle() {
        return title;
    }

    public BookDTO setTitle(String title) {
        this.title = title;
        return this;
    }

    public String getAuthor() {
        return author;
    }

    public BookDTO setAuthor(String author) {
        this.author = author;
        return this;
    }

    public List<String> getCategories() {
        return categories;
    }

    public BookDTO setCategories(List<String> categories) {
        this.categories = categories;
        return this;
    }

    public BookDetail getDetails() {
        return details;
    }

    public BookDTO setDetails(BookDetail details) {
        this.details = details;
        return this;
    }

    public String getTransientDescription() {
        return transientDescription;
    }

    public BookDTO setTransientDescription(String transientDescription) {
        this.transientDescription = transientDescription;
        return this;
    }

    public Date getCreationDate() {
        return creationDate;
    }

    public BookDTO setCreationDate(Date creationDate) {
        this.creationDate = creationDate;
        return this;
    }

    @Override
    public String toString() {
        return "BookDTO{" +
                "title='" + title + '\'' +
                ", author='" + author + '\'' +
                ", id='" + id + '\'' +
                ", transientDescription='" + transientDescription + '\'' +
                ", creationDate='" + creationDate + '\'' +
                ", categories=" + categories +
                ", details=" + details +
                '}';
    }
}

File: integration-tests/mongodb-panache-kotlin/src/test/kotlin/io/quarkus/it/mongodb/panache/BookDTO.kt
package io.quarkus.it.mongodb.panache

import com.fasterxml.jackson.annotation.JsonFormat
import io.quarkus.it.mongodb.panache.book.BookDetail
import java.util.ArrayList
import java.util.Date

/**
 * The IT uses a DTO and not directly the Book object because it should avoid the usage of ObjectId.
 */
class BookDTO {
    var title: String? = null
        private set

    var author: String? = null
        private set

    var id: String? = null
    var transientDescription: String? = null
        private set

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
    private var creationDate: Date? = null
    var categories: List<String> = ArrayList()
        private set

    var details: BookDetail? = null
        private set

    fun setTitle(title: String?): BookDTO {
        this.title = title
        return this
    }

    fun setAuthor(author: String?): BookDTO {
        this.author = author
        return this
    }

    fun setCategories(categories: List<String>): BookDTO {
        this.categories = categories
        return this
    }

    fun setDetails(details: BookDetail?): BookDTO {
        this.details = details
        return this
    }

    fun setTransientDescription(transientDescription: String?): BookDTO {
        this.transientDescription = transientDescription
        return this
    }

    fun getCreationDate(): Date? {
        return creationDate
    }

    fun setCreationDate(creationDate: Date?): BookDTO {
        this.creationDate = creationDate
        return this
    }

    @Override
    override fun toString(): String {
        return "BookDTO{" +
            "title='" +
            title +
            '\'' +
            ", author='" +
            author +
            '\'' +
            ", id='" +
            id +
            '\'' +
            ", transientDescription='" +
            transientDescription +
            '\'' +
            ", creationDate='" +
            creationDate +
            '\'' +
            ", categories=" +
            categories +
            ", details=" +
            details +
            '}'
    }
}
